File: mobile/test/core/pairing_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:nexusvpn/core/pairing.dart';

/// Pinned to the same cases as `client/agent/pairing_test.go`, because the
/// link is the entire interface between two devices that have never met: the
/// desktop writes it, this reads it, and neither can see the other's tests.
void main() {
  group('links the desktop produces', () {
    test('the standard form', () {
      final p = parsePairingLink(
        'nexusvpn://pair?s=http%3A%2F%2F192.168.1.20%3A8080&t=abc.def.ghi',
      );
      expect(p, isNotNull);
      expect(p!.serverUrl, 'http://192.168.1.20:8080');
      expect(p.token, 'abc.def.ghi');
    });

    test('the opaque form, which some QR readers produce', () {
      final p = parsePairingLink('nexusvpn:pair?s=http%3A%2F%2F10.0.0.5%3A8080&t=tok');
      expect(p?.serverUrl, 'http://10.0.0.5:8080');
      expect(p?.token, 'tok');
    });

    test('mixed case scheme, because scanners rewrite them', () {
      final p = parsePairingLink('NexusVPN://PAIR?s=http%3A%2F%2F10.0.0.5%3A8080&t=tok');
      expect(p?.serverUrl, 'http://10.0.0.5:8080');
    });

    test('surrounding whitespace from a paste', () {
      final p = parsePairingLink('  nexusvpn://pair?s=http%3A%2F%2F10.0.0.5%3A8080&t=tok  ');
      expect(p?.token, 'tok');
    });

    // Not hand-written: produced by running the same url.Values.Encode()
    // the desktop uses, with a real JWT shape in it. A parser tested only
    // against links its own author typed can pass while rejecting every link
    // the other half of the system actually emits.
    test('a link captured from the Go generator, verbatim', () {
      const generated =
          'nexusvpn://pair?s=http%3A%2F%2F192.168.1.20%3A8080'
          '&t=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1aWQiOiI3ZjNhIn0.sig';
      final p = parsePairingLink(generated);
      expect(p, isNotNull, reason: 'the phone cannot read what the desktop writes');
      expect(p!.serverUrl, 'http://192.168.1.20:8080');
      expect(
        p.token,
        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1aWQiOiI3ZjNhIn0.sig',
        reason: 'a JWT contains dots, which a sloppy split would eat',
      );
    });
  });

  group('what is not a pairing link', () {
    // An invite joins a network you are already signed in for; a pairing code
    // signs you in. Reading one as the other would claim against a token that
    // is not one.
    test('an invite link', () {
      expect(parsePairingLink('nexusvpn://join/ABC123'), isNull);
    });
    test('a code with no server cannot be redeemed anywhere', () {
      expect(parsePairingLink('nexusvpn://pair?t=tok'), isNull);
    });
    test('a server with no code is not an offer of anything', () {
      expect(parsePairingLink('nexusvpn://pair?s=http%3A%2F%2F10.0.0.5'), isNull);
    });
    test("somebody else's scheme", () {
      expect(parsePairingLink('otherapp://pair?s=http%3A%2F%2Fx&t=tok'), isNull);
    });
    test('an ordinary web address', () {
      expect(parsePairingLink('https://example.com/pair?s=x&t=y'), isNull);
    });
    test('empty', () {
      expect(parsePairingLink('   '), isNull);
    });
    test('a bare word', () {
      expect(parsePairingLink('ABC123'), isNull);
    });
  });
}

File: client/agent/pairing_test.go
package agent

import "testing"

// The link is the whole interface between two devices that have never met,
// so every shape it can arrive in is worth pinning.
func TestParsePairingLink(t *testing.T) {
	for _, tc := range []struct {
		name   string
		input  string
		server string
		token  string
		ok     bool
	}{
		{
			name:   "a link this app generated",
			input:  "nexusvpn://pair?s=http%3A%2F%2F192.168.1.20%3A8080&t=abc.def.ghi",
			server: "http://192.168.1.20:8080",
			token:  "abc.def.ghi",
			ok:     true,
		},
		{
			name:   "the opaque form, which some QR readers produce",
			input:  "nexusvpn:pair?s=http%3A%2F%2F10.0.0.5%3A8080&t=tok",
			server: "http://10.0.0.5:8080",
			token:  "tok",
			ok:     true,
		},
		{
			name:   "mixed case scheme, because scanners rewrite them",
			input:  "NexusVPN://PAIR?s=http%3A%2F%2F10.0.0.5%3A8080&t=tok",
			server: "http://10.0.0.5:8080",
			token:  "tok",
			ok:     true,
		},
		{
			name:   "surrounding whitespace from a paste",
			input:  "  nexusvpn://pair?s=http%3A%2F%2F10.0.0.5%3A8080&t=tok  ",
			server: "http://10.0.0.5:8080",
			token:  "tok",
			ok:     true,
		},

		// A join link carries a network invite and no session. Treating one
		// as the other would produce a claim against a token that is not one.
		{name: "an invite link is not a pairing link", input: "nexusvpn://join/ABC123"},

		{name: "no token", input: "nexusvpn://pair?s=http%3A%2F%2F10.0.0.5%3A8080"},
		{name: "no server", input: "nexusvpn://pair?t=tok"},
		{name: "someone else's scheme", input: "otherapp://pair?s=http%3A%2F%2Fx&t=tok"},
		{name: "empty", input: ""},
		{name: "not a link at all", input: "ABC123"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			server, token, ok := ParsePairingLink(tc.input)
			if ok != tc.ok {
				t.Fatalf("ok = %v, want %v", ok, tc.ok)
			}
			if server != tc.server || token != tc.token {
				t.Fatalf("got (%q, %q), want (%q, %q)", server, token, tc.server, tc.token)
			}
		})
	}
}

// TestReachableServerURLReplacesLoopback is the reason the link carries a
// server at all. The desktop app talks to its own control plane on
// 127.0.0.1, which on a phone means the phone.
func TestReachableServerURLReplacesLoopback(t *testing.T) {
	lan := lanAddress()
	if lan == "" {
		t.Skip("this machine has no routable address to substitute")
	}

	got := reachableServerURL("http://127.0.0.1:8080")
	if got == "http://127.0.0.1:8080" {
		t.Fatal("handed a phone the address of the phone")
	}
	want := "http://" + lan + ":8080"
	if got != want {
		t.Fatalf("got %q, want %q", got, want)
	}
}

// An address somebody deliberately configured is left alone: they know
// something about their network that this process does not.
func TestReachableServerURLLeavesRealAddressesAlone(t *testing.T) {
	for _, raw := range []string{
		"http://192.168.1.50:8080",
		"https://vpn.example.com",
		"http://10.1.2.3:9000",
	} {
		if got := reachableServerURL(raw); got != raw {
			t.Errorf("rewrote %q to %q", raw, got)
		}
	}
}

// A link is only useful if the thing that reads it is the thing that wrote
// it, so the two are checked against each other rather than separately.
func TestGeneratedLinksParseBack(t *testing.T) {
	for _, server := range []string{
		"http://192.168.1.20:8080",
		"http://10.0.0.5:8080",
		"https://vpn.example.com",
	} {
		link := "nexusvpn://pair?s=" + urlEscape(server) + "&t=some.jwt.value"
		gotServer, gotToken, ok := ParsePairingLink(link)
		if !ok {
			t.Fatalf("could not read back a link for %q", server)
		}
		if gotServer != server {
			t.Errorf("server round-tripped as %q, want %q", gotServer, server)
		}
		if gotToken != "some.jwt.value" {
			t.Errorf("token round-tripped as %q", gotToken)
		}
	}
}

func urlEscape(s string) string {
	out := ""
	for _, r := range s {
		switch {
		case r == ':' || r == '/' || r == '?' || r == '&' || r == '=':
			out += "%" + hexByte(byte(r))
		default:
			out += string(r)
		}
	}
	return out
}

func hexByte(b byte) string {
	const digits = "0123456789ABCDEF"
	return string([]byte{digits[b>>4], digits[b&0xf]})
}
